util/config.cc

```diff
@@ -1,90 +1,232 @@
-      subscriptions_endpoint_(ExpandEnvVars(subscriptions_endpoint)),
-      publish_endpoint_(ExpandEnvVars(publish_endpoint)),
-      validate_metrics_(validate_metrics),
-      check_cluster_endpoint_(ExpandEnvVars(check_cluster_endpoint)),
-      notify_alert_server_(notifyAlertServer),
-      publish_config_(std::move(publish_config)),
-      subscription_refresh_(subscription_refresh),
-      connect_timeout_(connect_timeout),
-      read_timeout_(read_timeout),
-      batch_size_(batch_size),
-      force_start_(force_start),
-      enable_main_(enable_main),
-      enable_subscriptions_(enable_subscriptions),
-      dump_metrics_(dump_metrics),
-      dump_subscriptions_(dump_subscriptions),
-      log_verbosity_(log_verbosity),
-      send_in_parallel_(send_in_parallel),
-      log_max_size_(log_max_size),
-      log_max_files_(log_max_files),
-      common_tags_(std::move(common_tags)) {}
-
 std::string Config::LoggingDirectory() const noexcept {
   return GetLoggingDir();
 }
 
 bool Config::IsMainEnabled() const noexcept {
   bool disabled = disabled_file_watcher_.exists();
-  return !disabled && enable_main_;
+  return !disabled && features.main;
 }
 
 bool Config::AreSubsEnabled() const noexcept {
   bool disabled = disabled_file_watcher_.exists();
-  return !disabled && enable_subscriptions_;
+  return !disabled && features.subscriptions;
+}
+
+static inline void put_if_nonempty(meter::Tags* tags, const char* key,
+                                   std::string&& val) {
+  if (!val.empty()) {
+    tags->add(key, val.c_str());
+  }
 }
 
-std::string ConfigToString(const Config& config) noexcept {
+static meter::Tags get_default_common_tags() {
+  meter::Tags common_tags_;
+  put_if_nonempty(&common_tags_, "nf.node", env::instance_id());
+  put_if_nonempty(&common_tags_, "nf.cluster", env::cluster());
+  put_if_nonempty(&common_tags_, "nf.app", env::app());
+  put_if_nonempty(&common_tags_, "nf.asg", env::asg());
+  put_if_nonempty(&common_tags_, "nf.stack", env::stack());
+  put_if_nonempty(&common_tags_, "nf.vmtype", env::vmtype());
+  put_if_nonempty(&common_tags_, "nf.task", env::taskid());
+  put_if_nonempty(&common_tags_, "nf.zone", env::zone());
+  put_if_nonempty(&common_tags_, "nf.region", env::region());
+  put_if_nonempty(&common_tags_, "nf.account", env::account());
+  return common_tags_;
+}
+
+Config::Config() noexcept
+    : disabled_file_watcher_(features.disabled_file),
+      common_tags_(get_default_common_tags()) {}
+
+std::unique_ptr<Config> Config::FromJson(const rapidjson::Document& json,
+                                         const Config& defaults) noexcept {
+  auto config = std::make_unique<Config>();
+
+  config->features = FeaturesConfig::FromJson(json, defaults.features);
+  config->http = HttpConfig::FromJson(json, defaults.http);
+  config->endpoints = EndpointConfig::FromJson(json, defaults.endpoints);
+  config->logs = LogConfig::FromJson(json, defaults.logs);
```

Comment:
I guess it has to be this way for backwards compatibility, but one downside is that it isn't clear which key belongs to what type of config. We could inadvertently use the same key from multiple configs and overload its meaning.